File: src/main/java/com/szokone/aem/gokotlin/java/filter/ComponentDecoratorProcessor.java
package com.szokone.aem.gokotlin.java.filter;

import com.szokone.aem.gokotlin.java.service.ConfigurableService;
import org.apache.commons.lang3.StringUtils;
import org.apache.sling.api.SlingHttpServletRequest;
import org.apache.sling.api.SlingHttpServletResponse;
import org.apache.sling.api.wrappers.SlingHttpServletResponseWrapper;

import javax.servlet.FilterChain;
import javax.servlet.ServletException;
import java.io.IOException;
import java.io.PrintWriter;
import java.io.StringWriter;

class ComponentDecoratorProcessor extends SlingHttpServletResponseWrapper {

    private final SlingHttpServletRequest request;

    private final SlingHttpServletResponse response;

    private final FilterChain chain;

    private final ConfigurableService service;

    private StringWriter sw;

    void process() throws IOException, ServletException {
        chain.doFilter(request, this);
        String html = sw != null ? sw.toString() : StringUtils.EMPTY;

        for (String search : service.getReplacements().keySet()) {
            if (html.contains(search)) {
                html = StringUtils.replaceAll(html, search, service.getReplacements().get(search));
            }
        }
        response.getWriter().write(html);
    }

    ComponentDecoratorProcessor(SlingHttpServletRequest request, SlingHttpServletResponse response, FilterChain chain, ConfigurableService service) {
        super(response);
        this.request = request;
        this.response = response;
        this.chain = chain;
        this.service = service;
    }

    @Override
    public PrintWriter getWriter() throws IOException {
        if (sw == null) {
            sw = new StringWriter();
        }
        return new PrintWriter(sw);
    }
}

File: src/main/kotlin/com/szokone/aem/gokotlin/kotlin/filter/ComponentDecoratorProcessor.kt
package com.szokone.aem.gokotlin.kotlin.filter

import com.szokone.aem.gokotlin.kotlin.service.ConfigurableService
import org.apache.commons.lang3.StringUtils
import org.apache.sling.api.SlingHttpServletRequest
import org.apache.sling.api.SlingHttpServletResponse
import org.apache.sling.api.wrappers.SlingHttpServletResponseWrapper
import java.io.IOException
import java.io.PrintWriter
import java.io.StringWriter
import javax.servlet.FilterChain

class ComponentDecoratorProcessor(private val request: SlingHttpServletRequest,
                                  private val response: SlingHttpServletResponse,
                                  private val chain: FilterChain?,
                                  private val service: ConfigurableService) : SlingHttpServletResponseWrapper(response) {

    private var sw: StringWriter? = null

    fun process() {
        chain?.doFilter(request, this)
        var html = if (sw != null) sw.toString() else StringUtils.EMPTY

        service.getReplacements().forEach {
            if (html.contains(it.component1())) html = StringUtils.replaceAll(html, it.component1(), it.component2())
        }
        response.writer.write(html)
    }


    @Throws(IOException::class)
    override fun getWriter(): PrintWriter {
        sw = sw ?: StringWriter()
        return PrintWriter(sw)
    }
}